### PR TITLE
DOCSP-30552: Collations guide

### PR DESCRIPTION
**Pull request by norareidy**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-rust/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-30552>
Staging - <https://preview-mongodbnorareidy.gatsbyjs.io/rust/DOCSP-30552-collations-guide/fundamentals/collations/>

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?

### CODE REVIEW

File: source/fundamentals/collations.txt [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+the driver compares ``length`` values as strings. Then, the code returns the following documents:
+
+.. code-block:: json
+   :copyable: false
+   
+   { "name" : "Emma", "length" : "474" }
+   { "name" : "Les Misérables", "length": "1462" }
+   { "name" : "Infinite Jest", "length" : "1104" }
+   { "name" : "Cryptonomicon", "length" : "918" }
+   { "name" : "Ça", "length" : "1138" }
```
Comment (rustagir) [inline source/fundamentals/collations.txt:315]: S: Instead of including the output, can you just give more detail in the preceding paragrah about what the matches are when they are compared as strings? For example, When the length values are compared as strings, 474 is considered alphabetically after 1000.

File: source/includes/fundamentals/code-snippets/collation.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        .collation(collation)
+        .build();
+
+    let index = IndexModel::builder()
+        .keys(doc! { "name": 1 })
+        .options(index_opts)
+        .build();
+
+    let result = my_coll.create_index(index, None).await?;
+
```
Comment (rustagir) [inline source/includes/fundamentals/code-snippets/collation.rs:71]: S: remove white space where unnecessary

**Review by rustagir — CHANGES_REQUESTED** [review]
a few things, happy to talk to fine tune anything

File: source/fundamentals/collations.txt [hunk 3]
```diff
@@ -0,0 +1,351 @@
+.. _rust-collations:
+
+==========
+Collations
+==========
+
```
Comment (rustagir) [inline source/fundamentals/collations.txt:6]: S: add meta tags!

File: source/fundamentals/collations.txt [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+Overview
+--------
+
+In this guide, you can learn how to use **collations** to order your find or
+aggregation operation results by string values. A collation is a set of character
+ordering conventions that apply to a specific language and locale.
```
Comment (rustagir) [inline source/fundamentals/collations.txt:18]: S: "that apply to" could be confusing here. Could you use "based on" or "that correspond to"?
The conventions actually apply to your data/results, so I think this sentence should be more clear.

File: source/fundamentals/collations.txt [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+
+This guide includes the following sections:
+
+- :ref:`rust-mongodb-collations` describes how MongoDB sorts strings according
+  to both the default collation and a custom collation
```
Comment (rustagir) [inline source/fundamentals/collations.txt:23]:
```suggestion
- :ref:`rust-mongodb-collations` describes how MongoDB sorts string values according
  to the default collation and custom collations
```

File: source/fundamentals/collations.txt [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+- :ref:`rust-specify-collation` describes how to create a ``Collation`` instance by
+  using builder functions
```
Comment (rustagir) [inline source/fundamentals/collations.txt:25]: S: remove some detail
```suggestion
- :ref:`rust-specify-collation` describes how to create a ``Collation`` struct instance
```

File: source/fundamentals/collations.txt [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+
+- :ref:`rust-mongodb-collations` describes how MongoDB sorts strings according
+  to both the default collation and a custom collation
+- :ref:`rust-specify-collation` describes how to create a ``Collation`` instance by
+  using builder functions 
+- :ref:`rust-collection-collation` describes how to set the collation for a
+  new collection
+- :ref:`rust-index-collation` describes how to set the collation for an index
+- :ref:`rust-op-collation` describes how to apply a collation to read, update, and
+  delete operations
```
Comment (rustagir) [inline source/fundamentals/collations.txt:30]:
```suggestion
- :ref:`rust-op-collation` describes how to apply a collation to CRUD operations
```

File: source/fundamentals/collations.txt [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+- :ref:`rust-collation-addtl-info` provides links to resources and API documentation
+  for types and methods mentioned in this guide
+
+.. _rust-mongodb-collations:
+
+MongoDB Collations
+------------------
+
+MongoDB sorts strings using *binary collation* by default. This collation
+method uses the `ASCII standard <https://en.wikipedia.org/wiki/ASCII>`_
```
Comment (rustagir) [inline source/fundamentals/collations.txt:40]: S: use a permanent link. Also, per the style guide, make link text clear. Perhaps you can move the wiki to a tip
```suggestion
method uses the `ASCII standard <https://en.wikipedia.org/wiki/ASCII>`_
```

File: source/fundamentals/collations.txt [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+
+.. code-block:: none
+
+   cote
+   coté
+   côte
+   côté
+
+In this sort order, "coté" is placed before "côte" because the ASCII standard "o" character value
+is lower than the "ô" character value.
```
Comment (rustagir) [inline source/fundamentals/collations.txt:64]: Q: what do you mean by "lower" here?

Comment (rustagir) [reply source/fundamentals/collations.txt:64]: Should this be reversed?

File: source/fundamentals/collations.txt [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+
+In this sort order, "coté" is placed after "côte" because Canadian French collation rules position
+the character "e" before the character "é".
+
+.. _rust-specify-collation:
+
+Specify a Collation
+-------------------
+
+You can define a collation by specifying your collation locale and related options in a
```
Comment (rustagir) [inline source/fundamentals/collations.txt:83]:
```suggestion
You can define a collation by specifying a collation locale and other options in a
```

File: source/fundamentals/collations.txt [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+``Collation`` struct instance. To begin building a ``Collation`` instance, call the
+``Collation::builder()`` method.
+
+.. note:: Instantiating Options
+   
+   The {+driver-short+} implements the Builder design pattern for the
+   creation of many different types, including ``Collation``. You can
+   use the ``builder()`` method to construct an instance of each type
+   by chaining option builder functions one at a time.
```
Comment (rustagir) [inline source/fundamentals/collations.txt:92]:
```suggestion
   by chaining option builder methods.
```

Comment (rustagir) [reply source/fundamentals/collations.txt:92]: use method for consistency with docs set

File: source/fundamentals/collations.txt [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+
+Call the following ``Collation::builder()`` functions to specify the collation. You must use
+the ``locale()`` function to build the ``Collation`` instance, but all other builder functions
+are optional:
```
Comment (rustagir) [inline source/fundamentals/collations.txt:96]:
```suggestion
The following table describes the builder methods that you can use to set fields of a ``Collation`` instance.
You must use the ``locale()`` function to build a valid ``Collation`` struct, but all other builder functions
are optional:
```

File: source/fundamentals/collations.txt [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+   :end-before: end-collation
+
+.. _rust-collection-collation:
+
+Set a Collation on a Collection
+-------------------------------
+
+When you create a new collection, you can define the collation for future operations
+called on that collection. Set the collation by using the ``collation()`` function
+to build a ``CreateCollectionOptions`` instance. Then, call the ``create_collection()``
```
Comment (rustagir) [inline source/fundamentals/collations.txt:172]:
```suggestion
called on that collection. Set the collation by using the ``collation()`` function
when creating a ``CreateCollectionOptions`` instance. Then, call the ``create_collection()``
```

File: source/fundamentals/collations.txt [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+method with your options instance as a parameter.
+
+.. _rust-create-collection:
+
+Create Collection with a Collation Example
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This example creates a collection called ``books`` and specifies a collation according
+to the ``"fr"``, or French, locale conventions. The ``strength`` field is set to
```
Comment (rustagir) [inline source/fundamentals/collations.txt:181]: S: reverse the two clauses of this setnece to alogn with the order of the code

File: source/fundamentals/collations.txt [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+``CollationStrength::Primary`` to ignore differences in diacritics.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/collation.rs
+   :language: rust
+   :dedent:
+   :start-after: start-create-collection
+   :end-before: end-create-collection
+
+Use the Collection's Collation Example
```
Comment (rustagir) [inline source/fundamentals/collations.txt:190]: S: right now, this title is a little hard to follow. Perhaps something like "Collation ordering demonstration" would be more apt? Happy to brainstorm!

File: source/fundamentals/collations.txt [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+      { "name": "Emma", "length": 474 }
+      { "name": "Cryptonomicon", "length": 918 }
+      { "name" : "Ça", "length" : "1138" }
+      
+
+If you don't specify a collation for the ``books`` collection, the ``find()`` method follows
+default binary collation rules to determine the ``name`` values that precede ``"Infinite Jest"``.
+These rules place words beginning with "Ç" after those beginning with "I". If the preceding find
+operation follows binary collation rules, the document that represents the book ``"Ça"`` is excluded
+from the results.
```
Comment (rustagir) [inline source/fundamentals/collations.txt:236]:
```suggestion
These rules place words beginning with "Ç" after those beginning with "I". So, when the preceding find
operation follows binary collation rules, the document with the ``name`` value of ``"Ça"`` does not match
the filter criteria.
```

File: source/fundamentals/collations.txt [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+
+.. _rust-index-collation:
+
+Set a Collation on an Index 
+---------------------------
+
+When you create a new index on a collection, you can define the collation for operations
+that use the index. To run an operation that uses the index and its collation, your operation
+and index must specify the same collation. Additionally, ensure that the index specifying
+the collation covers the query.
```
Comment (rustagir) [inline source/fundamentals/collations.txt:246]: S: omit the last sentence by integrating it with the first sentence
```suggestion
When you create a new index on a collection, you can define the collation for operations
that are covered by the index. To run an operation that uses the index and its collation, your operation
and index must specify the same collation.
```

File: source/fundamentals/collations.txt [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+
+Set the index collation by using the ``collation()`` function to build an ``IndexOptions`` instance.
+Then, pass your ``IndexOptions`` as an argument to an ``IndexModel`` builder function, and pass your
+``IndexModel`` as an argument to the ``create_index()`` method.
+
+Example
+~~~~~~~
+
+The following example uses the ``create_index()`` method to create an ascending index on the
+``name`` field and specifies a new collation with an ``"en_US"`` locale:
```
Comment (rustagir) [inline source/fundamentals/collations.txt:260]: S: avoid with
```suggestion
``name`` field and specifies a new collation corresponding to the ``"en_US"`` locale:
```

File: source/fundamentals/collations.txt [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+       Created index: name_1
+
+.. _rust-op-collation:
+
+Set a Collation on an Operation
+-------------------------------
+
+Operations that read, update, and delete documents from a collection can use collations.
+Applying a collation to an operation overrides any collation previously defined for a
+collection.
```
Comment (rustagir) [inline source/fundamentals/collations.txt:284]:
```suggestion
collection or index.
```

**Review by rustagir — CHANGES_REQUESTED** [review]
a few more things, but great job responding to comments!

File: source/fundamentals/collations.txt (hunk 9 above)
Comment (norareidy) [reply source/fundamentals/collations.txt:64]: I meant a lower ASCII code value, but yeah it's confusing so I will reword

File: source/fundamentals/collations.txt (hunk 15 above)
Comment (norareidy) [reply source/fundamentals/collations.txt]: I was a little hesitant about just titling it "Collation Ordering Demonstration" since that doesn't mention that it's using the collation set for the collection, but since it's a subheading under "Set a Collation on a Collection" I feel like it does work -- changed!

File: source/fundamentals/collations.txt [hunk 20]
```diff
@@ -0,0 +1,361 @@
...
+MongoDB sorts strings using *binary collation* by default. This collation method
+uses the ASCII standard character values to compare and order strings. Certain
+languages and locales have specific character ordering conventions that differ from
+the ASCII standard.
+
+.. tip::
+
+   To learn more about the ASCII standard, see the `ASCII
+   <https://en.wikipedia.org/w/index.php?title=ASCII&oldid=1180396712>`_ Wikipedia
+   page.
```
Comment (rustagir) [inline source/fundamentals/collations.txt:54]: S: there's a wiki role!
```suggestion
   To learn more about the ASCII standard, see the :wikipedia:`ASCII <w/index.php?title=ASCII&oldid=1180396712>` 
   Wikipedia page.
```

File: source/fundamentals/collations.txt [hunk 21]
```diff
@@ -0,0 +1,361 @@
...
+- côté
+
+When using the default binary collation, MongoDB sorts the words in the following order:
+
+.. code-block:: none
+
+   cote
+   coté
+   côte
+   côté
```
Comment (rustagir) [inline source/fundamentals/collations.txt:72]: S: set copyable false for these two blocks

**Review by rustagir — APPROVED** [review]
approved with two small suggestions, great job!

**Review by kevinAlbs — COMMENTED** [review]
LGTM
